pkg/controller/monitor/cnr.go

```diff
@@ -0,0 +1,336 @@
+		err = ctrl.emitCNRAnomalyMetric(cnr, reasonNumaExclusiveAnomaly)
+		if err != nil {
+			return err
+		}
+	}
+	// check numa allocatable sum anomaly
+	klog.Infof("Check Numa Allocatable Sum Anomaly...")
+	if ctrl.checkNumaAllocatableSumAnomaly(cnr) {
+		hasAnomaly = true
+		klog.Infof("Emit Numa Allocatable Sum Anomaly metric...")
+		err = ctrl.emitCNRAnomalyMetric(cnr, reasonNumaAllocatableSumAnomaly)
+		if err != nil {
+			return err
+		}
+	}
+
+	// check pod allocation sum anomaly
+	klog.Infof("Check Pod Allocation Sum Anomaly...")
+	if ctrl.checkPodAllocationSumAnomaly(cnr) {
+		hasAnomaly = true
+		klog.Infof("Emit Pod Allocation Sum Anomaly metric...")
+		err = ctrl.emitCNRAnomalyMetric(cnr, reasonPodAllocationSumAnomaly)
+		if err != nil {
+			return err
+		}
+	}
+
+	// if hasAnomaly is true, re-enqueue cnr use AddAfter func with 30s duration
+	if hasAnomaly {
+		klog.Infof("CNR %s has anomaly, re-enqueue it after 30s", cnr.Name)
+		ctrl.cnrSyncQueue.AddAfter(key, 30*time.Second)
+	}
+
+	return nil
+}
+
+// enqueueCNR enqueues the given CNR in the work queue.
+func (ctrl *CNRMonitorController) enqueueCNR(cnr *apis.CustomNodeResource) {
+	if cnr == nil {
+		klog.Warning("trying to enqueue a nil cnr")
+		return
+	}
+
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(cnr)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Cound't get key for CNR %+v: %v", cnr, err))
+		return
+	}
+	ctrl.cnrSyncQueue.Add(key)
+}
+
+func (ctrl *CNRMonitorController) addCNREventHandler(obj interface{}) {
+	klog.Infof("CNR create event found...")
+	cnr, ok := obj.(*apis.CustomNodeResource)
+	if !ok {
+		klog.Errorf("cannot convert obj to *apis.CNR")
+		return
+	}
+	klog.V(4).Infof("notice addition of cnr %s", cnr.Name)
+
+	ctrl.enqueueCNR(cnr)
+}
+
+func (ctrl *CNRMonitorController) updateCNREventHandler(_, newObj interface{}) {
+	klog.Infof("CNR update event found...")
+	cnr, ok := newObj.(*apis.CustomNodeResource)
+	if !ok {
+		klog.Errorf("cannot convert newObj to *apis.CNR")
+		return
+	}
+	klog.V(4).Infof("notice update of cnr %s", cnr.Name)
+
+	// check and emit cnr pod report lantency metric
+	klog.Infof("Check and Emit CNR Report Lantency metric...")
+	err := ctrl.checkAndEmitCNRReportLantencyMetric(cnr)
+	if err != nil {
+		klog.Errorf("check and emit cnr report lantency metric failed: %v", err)
+	}
+
+	ctrl.enqueueCNR(cnr)
+}
+
+func (ctrl *CNRMonitorController) updatePodEventHandler(oldObj, newObj interface{}) {
+	klog.Infof("Pod update event found...")
+	oldPod, ok := oldObj.(*corev1.Pod)
+	if !ok {
+		klog.Errorf("cannot convert oldObj to Pod")
+		return
+	}
+	newPod, ok := newObj.(*corev1.Pod)
+	if !ok {
+		klog.Errorf("cannot convert newObj to Pod")
+		return
+	}
+	if oldPod.Spec.NodeName == "" && newPod.Spec.NodeName != "" {
+		klog.Infof("notice pod: %v scheduled to node: %v", newPod.Name, newPod.Spec.NodeName)
+		// record pod scheduled time
+		ctrl.podTimeMap.Store(native.GenerateUniqObjectUIDKey(newPod), time.Now())
```

Comment:
If the start time and end time both use local time, will the difference be eliminated?

Reply:
Thanks for your suggestion, I was thinking if the slo we can tolerate is minute level, maybe the time accuracy requirement for reporting delay do not need very high?

Reply:
The point is not using `time.Now()` but when to call this `time.Now()`. In practice, the informer's resource handler or work queue's main loop method is not always executed at the time we expected. The reason is there are some locks and queues inside the informer implementation.
Indeed there is no big difference in most cases but we can't make sure the latency is just minute level, it might be longer than we expected especially in cluster which has enormous amount of pods.

Reply:
The precision of timestamps from Kubernetes Custom Resources (CR) is limited to the second level. We aim to achieve sub-second level measurement for CNR report latency. Since the behavior of pod scheduling and CNR reporting in the monitoring system is similar to the scheduler cache, measuring latency in this context is meaningful.